### PR TITLE
Image policy is resolving images on replica sets by default

### PR DESCRIPTION
**Pull request by smarterclayton**
Even when local names are not requested, the image policy plugin is
deciding to rewrite image references in replica sets that point to the
integrated registry (with tags) to use digests. This causes the
deployment controller that created them to get wedged (because it
detects a change to the template) and become unable to update status on
that replica set.

https://bugzilla.redhat.com/show_bug.cgi?id=1481801

### CODE REVIEW

**Comment by smarterclayton** [comment]
/retest

On Aug 18, 2017, at 8:55 PM, OpenShift CI Robot <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton>: The following tests
*failed*, say /retest to rerun them all:
Test name Commit Details Rerun command
ci/openshift-jenkins/extended_conformance_gce db9e95e
<https://github.com/openshift/origin/commit/db9e95e323424d5b3da67de26035692829890d8d>
link
<https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_gce/6268>
/test
extended_conformance_gce
ci/openshift-jenkins/extended_conformance_install_update 33aef8c
<https://github.com/openshift/origin/commit/33aef8cd036222d3b26af892e793cfb2a6d2de8a>
link
<https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_install_update/4433>
/test
extended_conformance_install_update

Full PR test history <https://openshift-gce-devel.appspot.com/pr/15868>. Your
PR dashboard <https://openshift-gce-devel.appspot.com/pr/smarterclayton>.
Please help us cut down on flakes by linking to
<https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
an open issue
<https://github.com/openshift/origin/issues?q=is:issue+is:open> when you
hit one in your PR.

Instructions for interacting with me using PR comments are available here
<https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>.
If you have questions or suggestions related to my behavior, please file an
issue against the kubernetes/test-infra
<https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
repository. I understand the commands that are listed here
<https://github.com/kubernetes/test-infra/blob/master/commands.md>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/15868#issuecomment-323489518>,
or mute
the thread

**Comment by smarterclayton** [comment]
We don't pipe through the full object today, but that is an option.  I do
think that initializers that mutate the target object will be important in
the future - we may just not have thought through this enough upstream
yet.  Direct hash checking has problems - perhaps mutating admission
controllers should have a way to signal that a deliberate mutation occurred
and collision should be avoided.

For now this is probably the simplest fix, and in 3.7 maybe you and I can
start a discussion with @kargakis and others about this

On Aug 19, 2017, at 5:45 AM, Tomáš Nožička <notifications@github.com> wrote:

*@tnozicka* commented on this pull request.

the other option would be to check if such object has a controllerRef set -
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/15868#pullrequestreview-57346608>,
or mute the thread

**Comment by tnozicka** [comment]
extended_conformance_install_update is blocked on https://github.com/openshift/origin/issues/15874

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 1]
```diff
@@ -585,6 +593,168 @@ func TestAdmissionResolveImages(t *testing.T) {
...
+						Strategy: buildapi.BuildStrategy{
+							CustomStrategy: &buildapi.CustomBuildStrategy{
+								From: kapi.ObjectReference{Kind: "ImageStreamTag", Name: "test:other"},
+							},
+						},
+					},
+				},
+			},
+		},
+		// does not rewrite the config because build has Attempt by default, which overrides global policy
```
Comment (tnozicka) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: Not `by default`, but explicitly set `Attempt`. (Default is `AttemptRewrite`.)

**Comment by tnozicka** [comment]
In future if this plugin would mutate only top level objects which can be found by checking for ownerRefs (which all controllers set) this kind of issues would go away I think. 

Like it would resolve the image only in Deployment itself and left RS and Pod with controllerRef set alone.

Unless there is a use case for mutating controlled objects.

**Comment by tnozicka** [comment]
In that case we need to work on supporting that upstream as you said and back it up with an use case / example.

**Comment by tnozicka** [comment]
network to GitHub failed
/test extended_conformance_install_update

**Comment by tnozicka** [comment]
new day; let's give it another shot
/test extended_conformance_install_update

**Comment by tnozicka** [comment]
@openshift/api-review needs API review :)

File: pkg/image/admission/imagepolicy/api/v1/defaults.go [hunk 2]
```diff
@@ -16,15 +16,19 @@ func SetDefaults_ImagePolicyConfig(obj *ImagePolicyConfig) {
 
 	if obj.ResolutionRules == nil {
 		obj.ResolutionRules = []ImageResolutionPolicyRule{
-			{TargetResource: GroupResource{Resource: "pods"}, LocalNames: true},
-			{TargetResource: GroupResource{Group: "build.openshift.io", Resource: "builds"}, LocalNames: true},
-			{TargetResource: GroupResource{Resource: "replicationcontrollers"}, LocalNames: true},
-			{TargetResource: GroupResource{Group: "extensions", Resource: "replicasets"}, LocalNames: true},
+			// pods are attempted for backwards compatibility.
+			{TargetResource: GroupResource{Resource: "pods"}, LocalNames: true, Type: Attempt},
```
Comment (smarterclayton) [inline pkg/image/admission/imagepolicy/api/v1/defaults.go]: Master rewriting is a bug, and that's caused by local namespace resolution.  If you look a few lines up the ResolveImage global (on the config) is defaulting to Attempt, and that's what we're controlling for here.  What went into 3.6 was too aggressive.

File: pkg/image/admission/imagepolicy/imagepolicy_test.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: I'm going to change the build, that was wrong.

**Comment by smarterclayton** [comment]
Updated to make builds Attempt instead of AttemptRewrite

**Comment by smarterclayton** [comment]
@liggitt need you to do quick API review.  3.6 is broken, in that rewrite should not happen unless requested.  Now, resolution rules override / more specific than the default resolution.

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 3]
```diff
@@ -585,6 +593,168 @@ func TestAdmissionResolveImages(t *testing.T) {
...
+						Strategy: buildapi.BuildStrategy{
+							SourceStrategy: &buildapi.SourceBuildStrategy{
+								From: kapi.ObjectReference{Kind: "DockerImage", Name: "integrated.registry/test/mysql@sha256:0000000000000000000000000000000000000000000000000000000000000001"},
+							},
+						},
+					},
+				},
+			},
+		},
+		// does not rewrite the config because build has DoNotAttempt by default, which overrides global policy
```
Comment (tnozicka) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: actually build has `Attempt` by default; the rest holds true

File: pkg/image/admission/imagepolicy/imagepolicy_test.go (hunk 1 above)
Comment (tnozicka) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: seems reasonable

File: pkg/image/admission/imagepolicy/api/v1/defaults.go [hunk 4]
```diff
@@ -16,15 +16,19 @@ func SetDefaults_ImagePolicyConfig(obj *ImagePolicyConfig) {
 
 	if obj.ResolutionRules == nil {
 		obj.ResolutionRules = []ImageResolutionPolicyRule{
-			{TargetResource: GroupResource{Resource: "pods"}, LocalNames: true},
-			{TargetResource: GroupResource{Group: "build.openshift.io", Resource: "builds"}, LocalNames: true},
-			{TargetResource: GroupResource{Resource: "replicationcontrollers"}, LocalNames: true},
-			{TargetResource: GroupResource{Group: "extensions", Resource: "replicasets"}, LocalNames: true},
+			// pods are rewrite attempted for backwards compatibility.
```
Comment (smarterclayton) [inline pkg/image/admission/imagepolicy/api/v1/defaults.go]: This is for resolution (do we look up this image value for possible application of rules).

**Comment by smarterclayton** [comment]
Updated with feedback from @liggitt 

Rename `Type` to `Policy`.  When defaulting ResolutionRules, default `Policy` to whatever the value of the top level `ResolveImages` value is, IF an execution rule is defined that references that resource (backwards compatible).

Makes `Policy` required so that users can't accidentally footgun themselves.

**Comment by smarterclayton** [comment]
/retest

Prometheus flake fixed (or reduced) in another PR

File: pkg/image/admission/imagepolicy/api/v1/defaults.go [hunk 5]
```diff
@@ -14,29 +14,49 @@ func SetDefaults_ImagePolicyConfig(obj *ImagePolicyConfig) {
...
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	scheme.AddTypeDefaultingFunc(&ImagePolicyConfig{}, func(obj interface{}) { SetDefaults_ImagePolicyConfig(obj.(*ImagePolicyConfig)) })
 	return nil
 }
+
+// executionRuleCoversResource returns true if gr is covered by rule.
+func executionRuleCoversResource(rule ImageExecutionPolicyRule, gr GroupResource) bool {
+	for _, target := range rule.OnResources {
+		if target.Group == gr.Group && (target.Resource == gr.Resource || target.Resource == "*") {
```
Comment (liggitt) [inline pkg/image/admission/imagepolicy/api/v1/defaults.go]: can group be `*`?

File: pkg/image/admission/imagepolicy/api/v1/defaults.go [hunk 6]
```diff
@@ -14,29 +14,49 @@ func SetDefaults_ImagePolicyConfig(obj *ImagePolicyConfig) {
...
-	for i := range obj.ExecutionRules {
-		if len(obj.ExecutionRules[i].OnResources) == 0 {
-			obj.ExecutionRules[i].OnResources = []GroupResource{{Resource: "pods", Group: kapi.GroupName}}
+		// default the resolution policy to the global default
+		for i := range obj.ResolutionRules {
+			if len(obj.ResolutionRules[i].Policy) == 0 {
+				obj.ResolutionRules[i].Policy = DoNotAttempt
+				for _, rule := range obj.ExecutionRules {
+					if executionRuleCoversResource(rule, obj.ResolutionRules[i].TargetResource) {
+						obj.ResolutionRules[i].Policy = obj.ResolveImages
```
Comment (liggitt) [inline pkg/image/admission/imagepolicy/api/v1/defaults.go]: nit: break

File: pkg/image/admission/imagepolicy/api/v1/defaults.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/image/admission/imagepolicy/api/v1/defaults.go]: No, no compelling reason.  apps could legitimately have a resource wildcard, but everything on the server seemed a stretch.

**Comment by mfojtik** [comment]
@smarterclayton this will require ansible update to set the resolutionRules ?

**Comment by liggitt** [comment]
API looks fine, just nits. make sure there's a release note about the new required field

**Comment by smarterclayton** [comment]
@mfojtik no we rely on defaulting today

Will fix the nits and repost in a short time.

**Comment by smarterclayton** [comment]
denitted

**Comment by tnozicka** [comment]
flake https://github.com/openshift/origin/issues/15977
/test extended_conformance_install_update

**Pull request merged**